airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java

```diff
@@ -0,0 +1,237 @@
+                  syncWorkflowInputs.getSyncInput(),
+                  connectionId);
+            } catch (final ChildWorkflowFailure childWorkflowFailure) {
+              if (!(childWorkflowFailure.getCause() instanceof CanceledFailure)) {
+                throw childWorkflowFailure;
+              }
+            }
+          }
+        });
+        syncWorkflowCancellationScope.run();
+      } catch (final CanceledFailure cf) {
+        // When a scope is cancelled temporal will thow a CanceledFailure as you can see here:
+        // https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/workflow/CancellationScope.java#L72
+        // The naming is very misleading, it is not a failure but the expected behavior...
+      }
+
+      if (workflowState.isUpdated()) {
+        log.error("A connection configuration has changed for the connection {}. The job will be restarted",
+            connectionUpdaterInput.getConnectionId());
+      } else if (workflowState.isDeleted()) {
+        // Stop the runs
+        final ConnectionDeletionInput connectionDeletionInput = new ConnectionDeletionInput(connectionUpdaterInput.getConnectionId());
+        connectionDeletionActivity.deleteConnection(connectionDeletionInput);
+        return;
+      } else if (workflowState.isCancelled()) {
+        jobCreationAndStatusUpdateActivity.jobCancelled(new JobCancelledInput(
+            maybeJobId.get()));
+      } else {
+        // report success
+        jobCreationAndStatusUpdateActivity.jobSuccess(new JobSuccessInput(
+            maybeJobId.get(),
+            maybeAttemptId.get()));
+
+        connectionUpdaterInput.setJobId(null);
+        connectionUpdaterInput.setAttemptNumber(1);
+        connectionUpdaterInput.setFromFailure(false);
+      }
+      continueAsNew(connectionUpdaterInput);
+    } catch (final Exception e) {
+      log.error("The connection update workflow has failed, will create a new attempt.", e);
+
+      jobCreationAndStatusUpdateActivity.attemptFailure(new AttemptFailureInput(
+          connectionUpdaterInput.getJobId(),
+          connectionUpdaterInput.getAttemptId()));
+
+      final int maxAttempt = configFetchActivity.getMaxAttempt().getMaxAttempt();
+      final int attemptNumber = connectionUpdaterInput.getAttemptNumber();
+
+      if (maxAttempt > attemptNumber) {
+        // restart from failure
+        connectionUpdaterInput.setAttemptNumber(attemptNumber + 1);
+        connectionUpdaterInput.setFromFailure(true);
+      } else {
+        jobCreationAndStatusUpdateActivity.jobFailure(new JobFailureInput(
+            connectionUpdaterInput.getJobId()));
+
+        Workflow.await(Duration.ofMinutes(1), () -> skipScheduling());
+
+        connectionUpdaterInput.setJobId(null);
+        connectionUpdaterInput.setAttemptNumber(1);
+        connectionUpdaterInput.setFromFailure(false);
+      }
+      continueAsNew(connectionUpdaterInput);
+    }
+  }
```

Comment:
```suggestion
    @Override
  public void run(final ConnectionUpdaterInput connectionUpdaterInput) throws RetryableException {
    try {
      if (connectionUpdaterInput.getWorkflowState() != null) {
        workflowState = connectionUpdaterInput.getWorkflowState();
      }
      try {
        syncWorkflowCancellationScope = Workflow.newCancellationScope(() -> {
          // Scheduling
          final ScheduleRetrieverInput scheduleRetrieverInput = new ScheduleRetrieverInput(connectionUpdaterInput.getConnectionId());
          final ScheduleRetrieverOutput scheduleRetrieverOutput = configFetchActivity.getTimeToWait(scheduleRetrieverInput);
          Workflow.await(scheduleRetrieverOutput.getTimeToWait(), () -> skipScheduling() || connectionUpdaterInput.isFromFailure());

          if (workflowState.isUpdated() || workflowState.isDeleted()) {
            return;
          }

          // Job and attempt creation
          maybeJobId = Optional.ofNullable(connectionUpdaterInput.getJobId()).or(() -> {
            final JobCreationOutput jobCreationOutput =
                jobCreationAndStatusUpdateActivity.createNewJob(new JobCreationInput(connectionUpdaterInput.getConnectionId()));
            return Optional.ofNullable(jobCreationOutput.getJobId());
          });

          maybeAttemptId = Optional.ofNullable(connectionUpdaterInput.getAttemptId()).or(() -> maybeJobId.map(jobId -> {
            final AttemptCreationOutput attemptCreationOutput = jobCreationAndStatusUpdateActivity.createNewAttempt(new AttemptCreationInput(jobId));
            return attemptCreationOutput.getAttemptId();
          }));

          // Sync workflow
          final SyncInput getSyncInputActivitySyncInput = new SyncInput(maybeAttemptId.get(), maybeJobId.get());
          final SyncOutput syncWorkflowInputs = getSyncInputActivity.getSyncWorkflowInput(getSyncInputActivitySyncInput);
          workflowState.setRunning(true);

          final SyncWorkflow childSync = Workflow.newChildWorkflowStub(SyncWorkflow.class,
              ChildWorkflowOptions.newBuilder()
                  .setWorkflowId("sync_" + maybeJobId.get())
                  .setTaskQueue(TemporalJobType.CONNECTION_UPDATER.name())
                  // This will cancel the child workflow when the parent is terminated
                  .setParentClosePolicy(ParentClosePolicy.PARENT_CLOSE_POLICY_TERMINATE)
                  .build());

          final UUID connectionId = connectionUpdaterInput.getConnectionId();
          try {
            childSync.run(
                syncWorkflowInputs.getJobRunConfig(),
                syncWorkflowInputs.getSourceLauncherConfig(),
                syncWorkflowInputs.getDestinationLauncherConfig(),
                syncWorkflowInputs.getSyncInput(),
                connectionId);
          } catch (final ChildWorkflowFailure childWorkflowFailure) {
            if (childWorkflowFailure.getCause() instanceof CanceledFailure) {
              return;
            }
            throw childWorkflowFailure;
          }
        });
        syncWorkflowCancellationScope.run();
      } catch (final CanceledFailure cf) {
        // When a scope is cancelled temporal will throw a CanceledFailure as you can see here:
        // https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/workflow/CancellationScope.java#L72
        // The naming is very misleading, it is not a failure but the expected behavior...
      }

      if (updateWorkflowAndCheckIsDeleted(connectionUpdaterInput)) {
        return;
      }
      
      continueAsNew(connectionUpdaterInput);
    } catch (final Exception e) {
      log.error("The connection update workflow has failed, will create a new attempt.", e);
      updateAttemptAndRetry(connectionUpdaterInput);
      continueAsNew(connectionUpdaterInput);
    }
  }

  private boolean updateWorkflowAndCheckIsDeleted(ConnectionUpdaterInput connectionUpdaterInput) {
    if (workflowState.isDeleted()) {
      // Stop the runs
      final ConnectionDeletionInput connectionDeletionInput = new ConnectionDeletionInput(connectionUpdaterInput.getConnectionId());
      connectionDeletionActivity.deleteConnection(connectionDeletionInput);
      return true;
    }

    if (workflowState.isUpdated()) {
      log.error("A connection configuration has changed for the connection {}. The job will be restarted",
          connectionUpdaterInput.getConnectionId());
    } else if (workflowState.isCancelled()) {
      jobCreationAndStatusUpdateActivity.jobCancelled(new JobCancelledInput(maybeJobId.get()));
    } else {
      // report success
      jobCreationAndStatusUpdateActivity.jobSuccess(new JobSuccessInput(maybeJobId.get(), maybeAttemptId.get()));
      connectionUpdaterInput.setJobId(null);
      connectionUpdaterInput.setAttemptNumber(1);
      connectionUpdaterInput.setFromFailure(false);
    }
    return false;
  }

  private void updateAttemptAndRetry(ConnectionUpdaterInput connectionUpdaterInput) {
    jobCreationAndStatusUpdateActivity.attemptFailure(
        new AttemptFailureInput(connectionUpdaterInput.getJobId(), connectionUpdaterInput.getAttemptId()));

    final int maxAttempt = configFetchActivity.getMaxAttempt().getMaxAttempt();
    final int attemptNumber = connectionUpdaterInput.getAttemptNumber();

    if (maxAttempt > attemptNumber) {
      // restart from failure
      connectionUpdaterInput.setAttemptNumber(attemptNumber + 1);
      connectionUpdaterInput.setFromFailure(true);
    } else {
      jobCreationAndStatusUpdateActivity.jobFailure(new JobFailureInput(connectionUpdaterInput.getJobId()));

      Workflow.await(Duration.ofMinutes(1), () -> skipScheduling());
      connectionUpdaterInput.setJobId(null);
      connectionUpdaterInput.setAttemptNumber(1);
      connectionUpdaterInput.setFromFailure(false);
    }
  }
```

Reply:
rough suggestion but I tried to:
1) exit early to avoid trailing right brackets and else conditions
2) pull out error and state handing logic into private methods for easier understanding

Reply:
I will try to do something similar, but returning fast might not be as easy as expected with the state being reset.